certbot-ci/certbot_integration_tests/utils/misc.py

```diff
@@ -30,16 +40,142 @@ def check_until_timeout(url):
+    copied = os.path.join(workspace, 'sample-config')
+    shutil.copytree(original, copied, symlinks=True)
+    return copied
+
+
+def read_certificate(cert_path):
+    with open(cert_path, 'r') as file:
+        data = file.read()
+
+    cert = crypto.load_certificate(crypto.FILETYPE_PEM, data.encode('utf-8'))
+    return crypto.dump_certificate(crypto.FILETYPE_TEXT, cert).decode('utf-8')
+
+
+class GraceFullTCPServer(socketserver.TCPServer):
+    allow_reuse_address = True
+
+
 @contextlib.contextmanager
-def execute_in_given_cwd(cwd):
-    """
-    Context manager that will execute any command in the given cwd after entering context,
-    and restore current cwd when context is destroyed.
-    :param str cwd: the path to use as the temporary current workspace for python execution
-    """
+def create_tcp_server(port):
     current_cwd = os.getcwd()
+    webroot = tempfile.mkdtemp()
+
+    def run():
+        GraceFullTCPServer(('', port), SimpleHTTPServer.SimpleHTTPRequestHandler).serve_forever()
+
+    process = multiprocessing.Process(target=run)
+
     try:
-        os.chdir(cwd)
-        yield
+        try:
+            os.chdir(webroot)
+            process.start()
+        finally:
+            os.chdir(current_cwd)
+
+        check_until_timeout('http://localhost:{0}/'.format(port))
+
+        yield webroot
     finally:
-        os.chdir(current_cwd)
+        try:
+            if process.is_alive():
+                process.terminate()
+                process.join()  # Block until process is effectively terminated
+        finally:
+            shutil.rmtree(webroot)
+
+
+def list_renewal_hooks_dirs(config_dir):
+    renewal_hooks_root = os.path.join(config_dir, 'renewal-hooks')
+    return [os.path.join(renewal_hooks_root, item) for item in ['pre', 'deploy', 'post']]
+
+
+def generate_test_file_hooks(config_dir, hook_probe):
+    if sys.platform == 'win32':
+        extension = 'bat'
+    else:
+        extension = 'sh'
+
+    renewal_hooks_dirs = list_renewal_hooks_dirs(config_dir)
+
+    for hook_dir in renewal_hooks_dirs:
+        try:
+            os.makedirs(hook_dir)
+        except OSError as error:
+            if error.errno != errno.EEXIST:
+                raise
+        hook_path = os.path.join(hook_dir, 'hook.{0}'.format(extension))
+        if extension == 'sh':
+            data = '''\
+#!/bin/bash -xe
+if [ "$0" == "{0}" ]; then
+    if [ -z "$RENEWED_DOMAINS" -o -z "$RENEWED_LINEAGE" ]; then
+        echo "Environment variables not properly set!" >&2
+        exit 1
+    fi
+fi
+echo $(basename $(dirname "$0")) >> "{1}"\
+'''.format(hook_path, hook_probe)
+        else:
+            data = '''\
+
+'''
+        with open(hook_path, 'w') as file:
+            file.write(data)
+        os.chmod(hook_path, os.stat(hook_path).st_mode | stat.S_IEXEC)
+
+
+def manual_http_hooks(http_server_root):
+    auth = (
+        '{0} -c "import os; '
+        "challenge_dir = os.path.join('{1}', '.well-known/acme-challenge'); "
```

Comment:
should probably be `os.path.join`ing the latter half of this as well.

Reply:
Done.